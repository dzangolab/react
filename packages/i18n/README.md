# Dzangolab React I18n

Adds internationalization to a react app.

## Requirements

* `@dzangolab/react-config: 0.1.0`

## Installation

```bash
pnpm install @dzangolab/react-config @dzangolab/react-i18n
```

## Usage

### Setup i18n

In `src/main.tsx`:

```ts
import i18n from "@dzangolab/react-i18n";
import i18n from "@dzangolab/react-i18n";
import React from "react";
import ReactDOM from "react-dom/client";

import App from "./App";
import config from "./config";


i18n(config.i18n);

...
```

### Resources

Create your resources under `src/locales/{locale}`.

For each locale, your resource should look like this:

```
{
  "locales": {
    "en": "English",
    "fr": "Français",
    "th": "ไทย"
  }
}
```

### Configuration

The package supports the same configuration options as the `react-i18next` package.

The configuration options must be defined under the `i18n` attribute of your app config (`AppConfig`).

```
# src/config.ts

const config: AppConfig = {
  ...
  i18n: {
    fallbackLng: "en",
    resources: {
      ...
    },
  },
  ...
};
```
