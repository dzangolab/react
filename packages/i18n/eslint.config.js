import path from "path";
import { fileURLToPath } from "url";

import { FlatCompat } from "@eslint/eslintrc";
import js from "@eslint/js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

const compat = new FlatCompat({
  baseDirectory: __dirname,
  recommendedConfig: js.configs.recommended,
  allConfig: js.configs.all,
});

export default [
  {
    ignores: ["dist", "node_modules"],
  },
  ...compat.extends("@12deg/eslint-config/react"),
];
