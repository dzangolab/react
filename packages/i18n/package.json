{
  "name": "@dzangolab/react-i18n",
  "version": "0.8.5",
  "description": "React I18n Plugin",
  "exports": {
    ".": {
      "import": "./dist/DzangolabReactI18n.es.js",
      "require": "./dist/DzangolabReactI18n.umd.js"
    },
    "./dist/DzangolabReactI18n.css": "./dist/style.css"
  },
  "main": "dist//DzangolabReactI18n.umd.js",
  "module": "dist//DzangolabReactI18n.es.js",
  "types": "./dist/src/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "vite build && tsc --emitDeclarationOnly",
    "lint": "eslint . --ext .js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts",
    "lint:fix": "eslint . --ext .js,.jsx,.cjs,.mjs,.ts,.tsx,.cts,.mts --fix",
    "sort-package": "npx sort-package-json",
    "test": "vitest --environment jsdom run --coverage",
    "test:component": "vitest --environment jsdom run component/",
    "test:snapshot": "vitest --environment jsdom run snapshot/",
    "test:unit": "vitest --environment jsdom run unit/",
    "test:watch": "vitest",
    "typecheck": "tsc --noEmit -p tsconfig.vitest.json --composite false"
  },
  "dependencies": {
    "@dzangolab/react-ui": "0.8.5",
    "i18next": "21.10.0",
    "i18next-browser-languagedetector": "6.1.8",
    "i18next-chained-backend": "^4.0.0",
    "i18next-resources-to-backend": "^1.0.0",
    "react-i18next": "11.18.6"
  },
  "devDependencies": {
    "@dzangolab/react-config": "0.8.5",
    "@testing-library/react": "14.0.0",
    "@types/i18n": "0.13.6",
    "@types/jsdom": "21.1.1",
    "@types/node": "18.11.18",
    "@types/react": "18.2.14",
    "@types/react-dom": "18.2.6",
    "@types/react-test-renderer": "18.0.0",
    "@vitejs/plugin-react": "3.1.0",
    "@vitest/coverage-c8": "0.30.1",
    "eslint": "8.42.0",
    "eslint-config-custom": "0.8.5",
    "eslint-plugin-react": "7.32.2",
    "prettier": "2.8.8",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "react-test-renderer": "18.2.0",
    "tsconfig": "0.8.5",
    "typescript": "5.1.3",
    "vite": "4.3.9",
    "vitest": "0.30.1"
  },
  "peerDependencies": {
    "@dzangolab/react-config": "0.8.5",
    "react": ">=18.2",
    "react-dom": ">=18.2"
  },
  "engines": {
    "node": ">=16",
    "pnpm": ">=8"
  }
}