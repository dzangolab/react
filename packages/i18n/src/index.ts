import { use } from "i18next";
import LanguageDetector from "i18next-browser-languagedetector";
import ChainedBackend from "i18next-chained-backend";
import resourcesToBackend from "i18next-resources-to-backend";
import { initReactI18next, useTranslation, Trans } from "react-i18next";

import LocaleSwitcher from "./components/LocaleSwitcher";

/* eslint-disable-next-line @typescript-eslint/no-unused-vars */
import type { AppConfig } from "@prefabs.tech/react-config";
import type { InitOptions } from "i18next";

const plugin = (options: InitOptions) => {
  use(LanguageDetector)
    .use(initReactI18next)
    .use(ChainedBackend)
    .init({
      ...options,
      interpolation: {
        escapeValue: false,
      },
      backend: {
        backends: [
          resourcesToBackend((language, namespace, callback) => {
            import(`./locales/${language}/${namespace}.json`)
              .then((resources) => {
                callback(null, resources);
              })
              .catch((error) => {
                callback(error, null);
              });
          }),
        ],
      },
    });
};

declare module "@prefabs.tech/react-config" {
  export interface AppConfig {
    i18n: InitOptions;
  }
}

export default plugin;

export { LocaleSwitcher, useTranslation, Trans };
