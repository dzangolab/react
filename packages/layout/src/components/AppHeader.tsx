import { LocaleSwitcher } from "@dzangolab/react-i18n";
import React, { useCallback, useState } from "react";

import Logo from "./Logo";
import MainMenu from "./MainMenu";
import useConfig from "../hooks/useConfig";

interface Properties {
  localeSwitcher?: React.ReactNode;
  logo?: React.ReactNode;
  logoRoute?: string;
  mainMenu?: React.ReactNode;
  mainMenuRoutes?: {
    name: string;
    route: string;
  }[];
  mainMenuOrientation?: "horizontal" | "vertical";
  navStyle?: "dropdown" | "left-slider";
  toggle?: React.ReactNode;
  userMenu?: React.ReactNode;
}

const AppHeader: React.FC<Properties> = (properties: Properties) => {
  const [expanded, setExpanded] = useState<boolean>(false);
  const { appName, layout: layoutConfig } = useConfig();

  const parseLogoAlt = useCallback(() => {
    if (!appName) {
      return;
    }

    return appName.replace("@", "").replace("/", " ");
  }, [appName]);

  const {
    localeSwitcher,
    logoRoute,
    logo,
    mainMenuOrientation,
    mainMenu,
    navStyle = "dropdown",
    toggle = <i className="pi pi-align-justify"></i>,
    userMenu,
    mainMenuRoutes,
  } = properties;

  return (
    <header aria-expanded={expanded} className={`${navStyle}`}>
      {logo || (
        <Logo
          src={layoutConfig?.logo}
          route={logoRoute || layoutConfig?.homeRoute}
          alt={layoutConfig?.logoAlt || parseLogoAlt()}
        />
      )}
      {mainMenu || (
        <MainMenu
          routes={mainMenuRoutes || layoutConfig?.mainMenu}
          orientation={mainMenuOrientation}
        />
      )}
      {userMenu}
      {localeSwitcher || <LocaleSwitcher />}
      <div className="toggle" onClick={() => setExpanded(!expanded)}>
        {toggle}
      </div>
    </header>
  );
};

export default AppHeader;
