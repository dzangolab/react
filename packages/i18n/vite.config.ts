import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

import react from "@vitejs/plugin-react";
import { defineConfig, loadEnv } from "vite";

import { dependencies, peerDependencies } from "./package.json";

// https://vitejs.dev/config/
export default defineConfig(({ mode }) => {
  process.env = { ...process.env, ...loadEnv(mode, process.cwd()) };

  return {
    build: {
      lib: {
        entry: resolve(dirname(fileURLToPath(import.meta.url)), "src/index.ts"),
        fileName: (format) => `DzangolabReactI18n.${format}.js`,
        name: "@dzangolab/react-i18n",
      },
      rollupOptions: {
        external: [
          ...Object.keys(peerDependencies),
          ...Object.keys(dependencies),
        ],
        output: {
          exports: "named",
          globals: {
            "@dzangolab/react-config": "DzangolabReactConfig",
            i18next: "I18next",
            "i18next-browser-languagedetector":
              "I18nextBrowserLanguagedetector",
            "i18next-chained-backend": "I18nextChainedBackend",
            "i18next-resources-to-backend": "I18nextResourcesToBackend",
            react: "React",
            "react-dom": "ReactDom",
            "react-i18next": "ReactI18next",
            "react-outside-click-handler": "ReactOutsideClickHandler",
          },
        },
      },
      target: "esnext",
      sourcemap: true,
    },
    plugins: [react()],
    resolve: {
      alias: {
        "@": fileURLToPath(new URL("./src", import.meta.url)),
      },
    },
    server: {
      port: Number(process.env.VITE_APP_PORT) || 8889,
    },
    optimizeDeps: {
      include: ["react/jsx-runtime"],
    },
  };
});
