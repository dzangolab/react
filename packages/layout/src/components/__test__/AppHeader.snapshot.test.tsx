import { AppConfig, configContext } from "@dzangolab/react-config";
import i18n from "@dzangolab/react-i18n";
import { create } from "react-test-renderer";
import { expect, test } from "vitest";

import AppHeader from "../AppHeader";

import type {
  ReactTestRenderer,
  ReactTestRendererJSON,
} from "react-test-renderer";

function toJson(component: ReactTestRenderer) {
  const result = component.toJSON();

  expect(result).toBeDefined();
  expect(result).not.toBeInstanceOf(Array);

  return result as ReactTestRendererJSON;
}

const appConfig: AppConfig = {
  apiBaseUrl: "/",
  app: {
    logo: "/logo.png",
    name: "Dzango Skeletons",
    version: "0.0.1-local",
  },
  appContext: "ADMIN",
  appPort: "20072",
  appTitle: "Dzango Skeletons",
  appVersion: "0.0.1",
  features: {
    showVersion: true,
  },
  i18n: {
    appendNamespaceToCIMode: true,
    debug: true,
    defaultNS: "app",
    fallbackLng: "en",
    supportedLngs: ["en", "fr"],
    react: {
      useSuspense: false,
    },
    resources: { en: {}, fr: {} },
  },
  locales: "en:English,fr:FranÃ§ais",
  localStorageNamespace: "__app_",
  websiteDomain: "//",
};

test("Component matches snapshot", () => {
  i18n(appConfig.i18n);

  const component = create(
    <configContext.Provider value={appConfig}>
      <AppHeader />
    </configContext.Provider>
  );

  const tree = toJson(component);
  expect(tree).toMatchSnapshot();
});
