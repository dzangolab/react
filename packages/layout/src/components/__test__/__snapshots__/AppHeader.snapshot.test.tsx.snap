// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component matches snapshot 1`] = `
<header>
  <div
    className="logo"
  >
    <a
      href="/"
    >
      <img
        alt="logo"
        src="/logo.png"
      />
    </a>
  </div>
  <div
    className="menu-container dropdown"
    data-aria-expanded={false}
  >
    <nav
      className="main-menu"
    >
      <div
        className="responsive-menu"
      >
        <ul>
          <li>
            <a
              href="/"
            >
              header.menu.home
            </a>
          </li>
          <li>
            <a
              href="/about"
            >
              header.menu.about
            </a>
          </li>
        </ul>
      </div>
    </nav>
    <div
      className="locale-switcher dropdown-container"
      data-aria-expanded={false}
      onClick={[Function]}
    >
      <div
        className="label-container"
      >
        <span
          className="label"
        >
          locales.en
        </span>
        <span
          className="icon"
        >
          â–¾
        </span>
      </div>
    </div>
  </div>
  <div
    className="toggle"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
          "height": "1.5rem",
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M32 96v64h448V96H32zm0 128v64h448v-64H32zm0 128v64h448v-64H32z"
      />
    </svg>
  </div>
</header>
`;
